### PR TITLE
More TF best practices. Add functionality to update bucket and read bucket metadata

### PR DESCRIPTION
**Pull request by ovaistariq**

### CODE REVIEW

File: internal/client.go [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign request: %v", err)
+	}
+
+	// Send the signed request using the wrapped http.Client
+	return c.httpClient.Do(req)
+}
+
+func (c *Client) bucketURL(bucketName string, queryParams map[string]string) string {
+	baseURL := fmt.Sprintf("%s/%s", c.endpoint, bucketName)
```
Comment (JigarJoshi) [inline internal/client.go:291]: this will work for all cases except the custom domain configured buckets.

File: internal/client.go [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+			return fmt.Errorf("failed to read request body: %v", err)
+		}
+		req.Body = io.NopCloser(bytes.NewReader(bodyBytes))
+	}
+
+	// Determine the payload hash (empty string if no payload)
+	payloadHash := hex.EncodeToString(sha256.New().Sum(bodyBytes))
+
+	// Sign the request using the signer
+	err := c.signer.SignHTTP(context.TODO(), c.credentials, req, payloadHash, "s3", c.cfg.Region, now)
```
Comment (JigarJoshi) [inline internal/client.go]: region for us is always going to be auto. maybe hardcode it?

File: internal/client.go (hunk 1 above)
Comment (ovaistariq) [reply internal/client.go:291]: This should work with custom domain buckets as well, as custom domains are just an additional way to access the bucket. Similar to how it works with S3 client for such buckets.

File: internal/client.go (hunk 2 above)
Comment (ovaistariq) [reply internal/client.go]: Good point. Let me make the change.

File: internal/client.go (hunk 1 above)
Comment (JigarJoshi) [reply internal/client.go:291]: bucket name is embedded in domain /<bucket-name> name will not work

Comment (ovaistariq) [reply internal/client.go:291]: Both will work: images.example.com, and fly.storage.tigris.dev/images.example.com

Comment (himank) [reply internal/client.go:291]: Yes, it will work. Even with custom domain the path prefix works.

File: internal/client.go [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+		return fmt.Errorf("update failed with error: %s", upResp.ErrorMessage)
+	}
+
+	return nil
+}
+
+func (c *Client) HeadBucket(ctx context.Context, bucketName string) (bool, error) {
+	_, err := c.s3Client.HeadBucket(ctx, &s3.HeadBucketInput{
+		Bucket: aws.String(bucketName),
+	}, withHeader(names.HeaderAmzIdentityId, c.credentials.AccessKeyID))
```
Comment (himank) [inline internal/client.go]: Do we need it here?

File: internal/names/names.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+	// Attributes for the terraform resources.
+	AttrBucket     = "bucket"
+	AttrAcl        = "acl"
+	AttrDomainName = "domain_name"
+
+	// Headers for the requests to Tigiris.
+	HeaderContentType   = "Content-Type"
+	HeaderAccept        = "Accept"
+	HeaderAmzContentSha = "X-Amz-Content-Sha256"
+	HeaderAmzIdentityId = "S3-Identity-Id"
```
Comment (himank) [inline internal/names/names.go]: We don't need to expose it.

File: internal/resource_bucket.go [hunk 5]
```diff
@@ -2,88 +2,190 @@ package internal
...
+	})
+
 	return nil
 }
 
-func resourceS3BucketUpdate(d *schema.ResourceData, meta interface{}) error {
-	// Since S3 buckets have limited update capabilities, this might be a no-op
-	return resourceS3BucketRead(d, meta)
+func resourceBucketUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	svc := meta.(*Client)
```
Comment (himank) [inline internal/resource_bucket.go]:
```
svc, ok := meta.(*Client)
if !ok {
    return ...
}
```

File: internal/types/types.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+type BucketMetadata struct {
+	Name             string         `json:"name"`
+	CacheControl     string         `json:"cache_control"`
+	ObjectRegions    string         `json:"object_regions"`
+	MD               *BucketMD      `json:"md"`
+	Shadow           *BucketShadow  `json:"shadow_bucket"`
+	Website          *BucketWebsite `json:"website"`
+	CreatedAt        time.Time      `json:"created_at"`
+	InitialCreatedAt time.Time      `json:"initial_created_at"`
```
Comment (himank) [inline internal/types/types.go]: Both `InitialCreatedAt` and `CreatedAt` are not exposed outside.

File: internal/client.go (hunk 3 above)
Comment (ovaistariq) [reply internal/client.go]: We need it here, otherwise the headbucket call fails?

File: internal/names/names.go (hunk 4 above)
Comment (ovaistariq) [reply internal/names/names.go]: Do you mean not export it from the package?

File: internal/types/types.go (hunk 6 above)
Comment (ovaistariq) [reply internal/types/types.go]: Ok, let me remove these fields

File: internal/names/names.go (hunk 4 above)
Comment (ovaistariq) [reply internal/names/names.go]: Updated

File: internal/client.go (hunk 3 above)
Comment (himank) [reply internal/client.go]: Okay, we can remove it once we rollout the fix in Gateway.

File: internal/names/names.go (hunk 4 above)
Comment (himank) [reply internal/names/names.go]: I mean we can completely remove it once we fix it in Gateway.

Comment (ovaistariq) [reply internal/names/names.go]: Yeah I will remove it then.

**Pull request merged**